selfservice/flow/saml/handler.go

```diff
@@ -0,0 +1,336 @@
+}
+
+// Handle /selfservice/methods/saml/metadata
+func (h *Handler) serveMetadata(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	config := h.d.Config()
+	if samlMiddleware == nil {
+		if err := h.instantiateMiddleware(r.Context(), *config); err != nil {
+			h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+		}
+	}
+
+	samlMiddleware.ServeMetadata(w, r)
+}
+
+// swagger:route GET /self-service/methods/saml/auth v0alpha2 initializeSelfServiceSamlFlowForBrowsers
+//
+// Initialize Authentication Flow for SAML (Either the login or the register)
+//
+// If you already have a session, it will redirect you to the main page.
+//
+//	Schemes: http, https
+//
+//	Responses:
+//	  200: selfServiceRegistrationFlow
+//	  400: jsonError
+//	  500: jsonError
+func (h *Handler) loginWithIdp(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	// Middleware is a singleton so we have to verify that it exists
+	if samlMiddleware == nil {
+		config := h.d.Config()
+		if err := h.instantiateMiddleware(r.Context(), *config); err != nil {
+			h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+		}
+	}
+
+	conf := h.d.Config()
+
+	// We have to get the SessionID from the cookie to inject it into the context to ensure continuity
+	cookie, err := r.Cookie(continuity.CookieName)
+	if err != nil {
+		h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+	}
+	body, _ := ioutil.ReadAll(r.Body)
+	r2 := r.Clone(context.WithValue(r.Context(), ory_kratos_continuity{}, cookie.Value))
+	r2.Body = ioutil.NopCloser(bytes.NewReader(body))
+	*r = *r2
+
+	// Checks if the user already have an active session
+	if e := new(session.ErrNoActiveSessionFound); errors.As(e, &e) {
+		// No session exists yet, we start the auth flow and create the session
+		samlMiddleware.HandleStartAuthFlow(w, r)
+	} else {
+		// A session already exist, we redirect to the main page
+		http.Redirect(w, r, conf.SelfServiceBrowserDefaultReturnTo(r.Context()).Path, http.StatusTemporaryRedirect)
+	}
+}
+
+func DestroyMiddlewareIfExists() {
+	if samlMiddleware != nil {
+		samlMiddleware = nil
+	}
+}
+
+func (h *Handler) instantiateMiddleware(ctx context.Context, config config.Config) error {
+	// Create a SAMLProvider object from the config file
+	var c samlstrategy.ConfigurationCollection
+	conf := config.SelfServiceStrategy(ctx, "saml").Config
+	if err := jsonx.
+		NewStrictDecoder(bytes.NewBuffer(conf)).
+		Decode(&c); err != nil {
+		return errors.Wrapf(err, "Unable to decode config %v", string(conf))
+	}
+
+	// Key pair to encrypt and sign SAML requests
+	keyPair, err := tls.LoadX509KeyPair(strings.Replace(c.SAMLProviders[len(c.SAMLProviders)-1].PublicCertPath, "file://", "", 1), strings.Replace(c.SAMLProviders[len(c.SAMLProviders)-1].PrivateKeyPath, "file://", "", 1))
+	if err != nil {
+		return err
+	}
+	keyPair.Leaf, err = x509.ParseCertificate(keyPair.Certificate[0])
+	if err != nil {
+		return err
```

Comment:
Please wrap errors in descriptive error messages using the `herodot` package (e.g. `herodot.ErrBadRequest`) and use `errors.WithStack()` to include stack traces